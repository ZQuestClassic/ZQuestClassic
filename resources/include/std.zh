// std.zh
// 28th December, 2020 for ZC 2.55 Alpha 86
// Standard header for ZScript.
// Please don't edit this file. Instead, create a new header file for your
// quests' scripts,  add constants and procedures to that file,
// and #include it separately.

//Options
#option BINARY_32BIT off
	//Would break a number of things if set on
#option SHORT_CIRCUIT on
	//Should work for everything in std_zh, and just speed it up

#includepath "include"

//Version Metadata
#include "std_zh/std_meta.zh"

//User Configuration Files
#include "std_zh/std_user_defs.zh" //Typedefs: Always first.
#include "std_zh/std.cfg"

//Main Files
#include "std_zh/limits.zh"
#include "std_zh/std_constants.zh"
#include "std_zh/ffrules.zh"
#include "std_zh/std_functions.zh"
#include "std_zh/math.zh"
#include "std_zh/weapon.zh"
#include "std_zh/std_sideview.zh" //Sideview and platforming

//ZScript string handling functions, akin to those found in C
//#include "string.zh"
#include "std_zh/string/string_constants.zh" //Now #includeed from here. This prevents conflicts for those people
#include "std_zh/string/string_functions.zh" //who would #include string.zh from a secondary source, as that file is now 
			     //empty, and the string.zh components load from here. 

//Keyboard handling for 2.54 and above
#include "std_zh/std_keyboard.zh"

//Locale
#include "std_zh/locale/std_EnglishIntl.zh" //International English Localisation

//Legacy Wrappers
#include "std_zh/std_legacy.zh"