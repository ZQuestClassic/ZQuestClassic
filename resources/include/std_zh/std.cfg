//Settings

const int STD_FORCE_OLD_DISTANCE = 1;	//If enabled, Distance() will use Sqrt() instead of SafeSqrt().
					//! If you use this, disable the quest rule 'Log Script Errors to Allegro.log
					//! before distributing your quest!
					
const int STD_DISTANCE_USE_SAFESQRT = 0; 	//If enebaled, Distance will use SafeSqrt(). Otherwise, it will return a factor, 
					
const int STD_EMULATE_ENGINE_COLLISION = 0; // If enabled, Collision hitboxs will more precisely emulate the
					    // internal collision. This differs to how scripted collision
					    // has always worked.
					    

const int STD_CENTERXY_USES_HITBOXES = 0; // If enabled, CenterX() and CenterY() will use object hitboxes
					  // and offsets by default, rather than sprite sizes when calculating
					  // the centre of objects.
					  

const int STD_NEVER_USE_SAFESQRT = 1; //Enable to force any calls to SafeSqrt() to use Sqrt() instead.

const int MAX_HIGHEST_LEVEL_ITEM_CHECK = 255; //The highest item ID checked by GetHighestLevelItem(). 

const int REVISED_ROTATEPOINT_X_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointX()
const int REVISED_ROTATEPOINT_Y_FUNCTION = 1; //Set to 0 to use 2.50.x RotatePointY()


const int MAX_NOACTION_INPUT = 14; //Checks A, B, L, R, U/D/L/R, Ex1, Ex2, Ex3, Ex4
	//Use CB_MAX to force the analogue stick false. 
const int NOACTION_SKIP_MAP = 0; //If enabled, NoAction will not disable pressing the Map button
const int NOACTION_SKIP_START = 0; //If enabled, NoAction will not disable pressing the Start button

const int OLD_ONSIDEPLATFORM = 0; //Change to 1 to use old OnSidePlatform calls.

const int OLD_RAND_FUNCS = 0; //Change to 1 to use old Rand() logic
