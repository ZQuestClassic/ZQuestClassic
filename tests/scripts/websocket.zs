#include "std.zh"

generic script WebSocketScript
{
    void run()
    {
        printf("Connecting to websocket.\n");
        websocket ws = Game->LoadWebSocket("ws://ws.ifelse.io");
        ws->Own();
        bool hasSentMessage = false;

        while (true)
        {
            if (ws->State == WEBSOCKET_STATE_CONNECTING)
            {
                Waitframe();
                continue;
            }
            else if (ws->State == WEBSOCKET_STATE_CLOSED)
            {
                char32 error[0];
                ws->GetError(error);
                printf("Failed to connect: %s\n", error);
                break;
            }

            if (ws->State != WEBSOCKET_STATE_OPEN)
            {
                printf("Failed to connect: %l\n", ws->State);
                break;
            }

            if (!hasSentMessage)
            {
                printf("Connected!\n");
                ws->Send("hello world!");
                ws->Send("domo arigato!", WEBSOCKET_MESSAGE_TYPE_BINARY);
                hasSentMessage = true;
            }

            while (ws->HasMessage)
            {
                int message_ptr = ws->Receive();
                int type = ws->MessageType;
                if (type == WEBSOCKET_MESSAGE_TYPE_TEXT)
                    printf("got text message: %s\n", message_ptr);
                else if (type == WEBSOCKET_MESSAGE_TYPE_BINARY)
                {
                    int len = SizeOfArray(message_ptr);
                    printf("got binary message of length: %d\n", len);
                    for (int i = 0; i < len; ++i)
                        printf("%d: %l\n", i, message_ptr[i]);
                }
                else
                    printf("got message, but with unknown type: %l\n", type);
            }

            Waitframe();
        }
    }
}
