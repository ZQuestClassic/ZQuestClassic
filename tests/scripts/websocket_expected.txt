Compiling 'websocket.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Success!
Compile finished with exit code '0' (success)

PUSHR REFGENERICDATA; void run() Body Start
PUSHARGSV 0,0.0010
SETR d4,SP2
ALLOCATEMEMV d2,26
STORE d2,0.0009
WRITEPODSTRING d2,"Connecting to websocket.\n"
LOAD d2,0.0009; InlineFunc[void printf(char32)] Params Start
PUSHR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0009
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,18
STORE d2,0.0008
WRITEPODSTRING d2,"ws://ws.ifelse.io"
LOAD d2,0.0008; InlineFunc[WebSocket Game->LoadWebSocket(char32)] Params
WEBSOCKET_LOAD d2; InlineFunc[WebSocket Game->LoadWebSocket(char32)] Body
LOAD d3,0.0008
DEALLOCATEMEMR d3
STORE d2,0.0007
LOAD d2,0.0007; InlineFunc[void WebSocket->Own()] Params Start
SETR REFWEBSOCKET,d2; InlineFunc[void WebSocket->Own()] Params End
WEBSOCKET_OWN; InlineFunc[void WebSocket->Own()] Body
STOREV 0,0.0006
LOAD d2,0.0007; if() #189 Test AND while(true) #8 Body [Opt:AlwaysOn] Start
PUSHR d2
POP REFWEBSOCKET
SETR d2,WEBSOCKET_STATE
COMPAREV d2,0
SETCMP d2,I==
COMPAREV d2,0
GOTOCMP 36,==; Test 'if'
WAITFRAME; InlineFunc[void Waitframe()] Body AND if() #189 Body Start
GOTO 25; continue 1;
GOTO 67; if() #189 Body End
LOAD d2,0.0007; if() #558 Test AND if() #189 Else Start
PUSHR d2
POP REFWEBSOCKET
SETR d2,WEBSOCKET_STATE
COMPAREV d2,0.0003
SETCMP d2,I==
COMPAREV d2,0
GOTOCMP 67,==; Test 'if'
ALLOCATEMEMV d2,0; if() #558 Body Start
STORE d2,0.0005
LOAD d2,0.0007; InlineFunc[void WebSocket->GetError(char32)] Params Start
PUSHR d2
LOAD d2,0.0005; InlineFunc[void WebSocket->GetError(char32)] Params End
POP REFWEBSOCKET; InlineFunc[void WebSocket->GetError(char32)] Body Start
WEBSOCKET_ERROR d2; InlineFunc[void WebSocket->GetError(char32)] Body End
ALLOCATEMEMV d2,23
STORE d2,0.0004
WRITEPODSTRING d2,"Failed to connect: %s\n"
LOAD d2,0.0004; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0005
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0004
DEALLOCATEMEMR d3
LOAD d3,0.0005
DEALLOCATEMEMR d3
GOTO 225; break 1;
LOAD d3,0.0005
DEALLOCATEMEMR d3; if() #558 Body End
LOAD d2,0.0007; if() #189 Else End AND if() #559 Test
PUSHR d2
POP REFWEBSOCKET
SETR d2,WEBSOCKET_STATE
COMPAREV d2,0.0001
SETCMP d2,I!=
COMPAREV d2,0
GOTOCMP 89,==; Test 'if'
ALLOCATEMEMV d2,23; if() #559 Body Start
STORE d2,0.0005
WRITEPODSTRING d2,"Failed to connect: %l\n"
LOAD d2,0.0005; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0007
PUSHR d2
POP REFWEBSOCKET
PUSHVARGR WEBSOCKET_STATE; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0005
DEALLOCATEMEMR d3
GOTO 225; break 1; AND if() #559 Body End
LOAD d2,0.0006; unless() #560 Test
COMPAREV d2,0
GOTOCMP 129,!=; Test 'unless'
ALLOCATEMEMV d2,12; unless() #560 Body Start
STORE d2,0.0005
WRITEPODSTRING d2,"Connected!\n"
LOAD d2,0.0005; InlineFunc[void printf(char32)] Params Start
PUSHR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0005
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,13
STORE d2,0.0004
WRITEPODSTRING d2,"hello world!"
LOAD d2,0.0007; InlineFunc[void WebSocket->Send(char32, long)] Params Start
PUSHR d2
LOAD d2,0.0004
PUSHR d2
SETV d2,0.0001; InlineFunc[void WebSocket->Send(char32, long)] Params End
POP d3; InlineFunc[void WebSocket->Send(char32, long)] Body Start
POP REFWEBSOCKET
WEBSOCKET_SEND d2,d3; InlineFunc[void WebSocket->Send(char32, long)] Body End
LOAD d3,0.0004
DEALLOCATEMEMR d3
ALLOCATEMEMV d2,14
STORE d2,0.0003
WRITEPODSTRING d2,"domo arigato!"
LOAD d2,0.0007; InlineFunc[void WebSocket->Send(char32, long)] Params Start
PUSHR d2
LOAD d2,0.0003
PUSHR d2
SETV d2,0.0002; InlineFunc[void WebSocket->Send(char32, long)] Params End
POP d3; InlineFunc[void WebSocket->Send(char32, long)] Body Start
POP REFWEBSOCKET
WEBSOCKET_SEND d2,d3; InlineFunc[void WebSocket->Send(char32, long)] Body End
LOAD d3,0.0003
DEALLOCATEMEMR d3
SETV d2,1
STORE d2,0.0006; unless() #560 Body End
LOAD d2,0.0007; while() #9 Test
PUSHR d2
POP REFWEBSOCKET
SETR d2,WEBSOCKET_HAS_MESSAGE
COMPAREV d2,0
GOTOCMP 223,==; Test 'while'
LOAD d2,0.0007; InlineFunc[long WebSocket->Receive()] Params Start AND while() #9 Body Start
SETR REFWEBSOCKET,d2; InlineFunc[long WebSocket->Receive()] Params End
WEBSOCKET_RECEIVE d2; InlineFunc[long WebSocket->Receive()] Body
STORE d2,0.0005
LOAD d2,0.0007
PUSHR d2
POP REFWEBSOCKET
SETR d2,WEBSOCKET_MESSAGE_TYPE
STORE d2,0.0004
LOAD d2,0.0004; if() #190 Test
COMPAREV d2,0.0001
SETCMP d2,I==
COMPAREV d2,0
GOTOCMP 161,==; Test 'if'
ALLOCATEMEMV d2,22; if() #190 Body Start
STORE d2,0.0003
WRITEPODSTRING d2,"got text message: %s\n"
LOAD d2,0.0003; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0005
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0003
DEALLOCATEMEMR d3
GOTO 129; if() #190 Body End
LOAD d2,0.0004; if() #191 Test AND if() #190 Else Start
COMPAREV d2,0.0002
SETCMP d2,I==
COMPAREV d2,0
GOTOCMP 211,==; Test 'if'
LOAD d2,0.0005; InlineFunc[int SizeOfArray(untyped)] Params AND if() #191 Body Start
ARRAYSIZE d2; InlineFunc[int SizeOfArray(untyped)] Body
STORE d2,0.0003
ALLOCATEMEMV d2,34
STORE d2,0.0002
WRITEPODSTRING d2,"got binary message of length: %d\n"
LOAD d2,0.0002; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0003
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0002
DEALLOCATEMEMR d3
STOREV 0,0.0001; for() #187 setup
LOAD d2,0.0001; for() #187 LoopTest
PUSHR d2
LOAD d2,0.0003
POP d3
COMPARER d3,d2
SETCMP d2,I<
COMPAREV d2,0
GOTOCMP 129,==; for() #187 TestFail?
ALLOCATEMEMV d2,8; for() #187 Body Start
STORE d2,0
WRITEPODSTRING d2,"%d: %l\n"
LOAD d2,0; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0001
PUSHVARGR d2
LOAD d2,0.0005
PUSHR d2
LOAD d2,0.0001
POP d0
READPODARRAYR d2,d2
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0
DEALLOCATEMEMR d3; for() #187 Body End
LOAD d2,0.0001; for() #187 LoopIncrement
ADDV d2,1
STORE d2,0.0001
GOTO 181; for() #187 End
GOTO 129; if() #191 Body End
ALLOCATEMEMV d2,40; if() #191 Else Start
STORE d2,0.0003
WRITEPODSTRING d2,"got message, but with unknown type: %l\n"
LOAD d2,0.0003; InlineFunc[void printf(char32)] Params Start
PUSHR d2
LOAD d2,0.0004
PUSHVARGR d2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP d5; InlineFunc[void printf(char32)] Body End
LOAD d3,0.0003
DEALLOCATEMEMR d3; if() #191 Else End
GOTO 129; if() #190 Else End AND while() #9 Body End AND while() #9 End
WAITFRAME; InlineFunc[void Waitframe()] Body AND while(true) #8 Body [Opt:AlwaysOn] End
GOTO 25; while() #8 End
POPARGS d5,0.0011
QUIT; void run() Body End


ALLOCATEGMEMV d2,4096; void run() Body Start
SETR gd1,d2
QUIT; void run() Body End